File: src/main/java/com/github/unchama/seichiassist/commands/RegionOwnerTransferCommand.kt
package com.github.unchama.seichiassist.commands

import com.github.unchama.contextualexecutor.asNonBlockingTabExecutor
import com.github.unchama.contextualexecutor.builder.ArgumentParserScope.ScopeProvider.parser
import com.github.unchama.contextualexecutor.builder.Parsers
import com.github.unchama.effect.TargetedEffect
import com.github.unchama.effect.asMessageEffect
import com.github.unchama.seichiassist.commands.contextual.builder.BuilderTemplates.playerCommandBuilder
import com.sk89q.worldguard.bukkit.WorldGuardPlugin
import com.sk89q.worldguard.protection.regions.ProtectedRegion
import org.bukkit.Bukkit
import org.bukkit.entity.Player

object RegionOwnerTransferCommand {
  @Suppress("RedundantSuspendModifier")
  private suspend fun attemptRegionTransfer(donner: Player, recipient: Player, region: ProtectedRegion): TargetedEffect<Player> {
    val owners = region.owners

    if (!owners.contains(donner.uniqueId)) {
      return "オーナーではないため権限を譲渡できません。".asMessageEffect()
    }

    if (owners.size() != 1) {
      return "オーナーが複数人いるため権限を譲渡できません。".asMessageEffect()
    }

    owners.clear()
    owners.addPlayer(recipient.uniqueId)

    return "${recipient.name}に${region.id}のオーナー権限を譲渡しました。".asMessageEffect()
  }

  val executor = playerCommandBuilder
      .argumentsParsers(listOf(
          Parsers.identity,
          parser { recipientName ->
            val recipient = Bukkit.getPlayer(recipientName)

            if (recipient != null) {
              succeedWith(recipient)
            } else {
              failWith("${recipientName}というプレイヤーはサーバーに参加したことがありません。")
            }
          }
      ))
      .execution { context ->
        val regionName = context.args.parsed[0] as String
        val newOwner = context.args.parsed[1] as Player

        val sender = context.sender

        val region = WorldGuardPlugin.inst().getRegionManager(sender.world).getRegion(regionName)
            ?: return@execution "${regionName}という名前の保護は存在しません。".asMessageEffect()

        attemptRegionTransfer(sender, newOwner, region)
      }
      .build()
      .asNonBlockingTabExecutor()
}

File: src/main/java/com/github/unchama/seichiassist/commands/RegionOwnerTransferCommand.scala
package com.github.unchama.seichiassist.commands

import cats.effect.IO
import com.github.unchama.contextualexecutor.builder.{ArgumentParserScope, Parsers}
import com.github.unchama.seichiassist.commands.contextual.builder.BuilderTemplates.playerCommandBuilder
import com.github.unchama.targetedeffect.MessageEffects._
import com.github.unchama.targetedeffect.TargetedEffect.TargetedEffect
import com.github.unchama.util.kotlin2scala.SuspendingMethod
import com.sk89q.worldguard.bukkit.WorldGuardPlugin
import com.sk89q.worldguard.protection.regions.ProtectedRegion
import org.bukkit.Bukkit
import org.bukkit.command.TabExecutor
import org.bukkit.entity.Player

object RegionOwnerTransferCommand {
  @SuspendingMethod
  private def attemptRegionTransfer(donner: Player, recipient: Player, region: ProtectedRegion): TargetedEffect[Player] = {
    val owners = region.getOwners

    if (!owners.contains(donner.getUniqueId)) {
      return "オーナーではないため権限を譲渡できません。".asMessageEffect()
    }

    if (owners.size() != 1) {
      return "オーナーが複数人いるため権限を譲渡できません。".asMessageEffect()
    }

    owners.clear()
    owners.addPlayer(recipient.getUniqueId)

    s"${recipient.getName}に${region.getId}のオーナー権限を譲渡しました。".asMessageEffect()
  }

  import ArgumentParserScope._
  val executor: TabExecutor = playerCommandBuilder
    .argumentsParsers(List(
      Parsers.identity,
      recipientName => {
        Bukkit.getPlayer(recipientName) match {
          case recipient: Player => succeedWith(recipient)
          case _ => failWith(s"${recipientName}というプレイヤーはサーバーに参加したことがありません。")
        }
      }
    ))
    .execution { context =>
      val regionName = context.args.parsed[0].asInstanceOf[String]
      val newOwner = context.args.parsed[1].asInstanceOf[Player]

      val sender = context.sender

      val region = WorldGuardPlugin.inst().getRegionManager(sender.getWorld).getRegion(regionName)
      if (region == null) {
        IO(s"${regionName}という名前の保護は存在しません。".asMessageEffect())
      }

      IO(attemptRegionTransfer(sender, newOwner, region))
    }
    .build()
    .asNonBlockingTabExecutor()
}